actors/market/src/lib.rs

```diff
@@ -543,137 +550,226 @@ impl Actor {
-                    log::warn!(
-                        "failed to activate sector containing duplicate deals {:?}",
-                        p.deal_ids
-                    );
-                    batch_gen.add_fail(ExitCode::USR_ILLEGAL_ARGUMENT);
-                    continue;
-                }
+            let mut sectors_deals: Vec<(SectorNumber, SectorDealIDs)> = vec![];
 
-                let proposals = match get_proposals(&proposal_array, &p.deal_ids, st.next_id) {
-                    Ok(proposals) => proposals,
-                    Err(e) => {
-                        log::warn!("failed to get proposals for deals {:?}: {:?}", p.deal_ids, e);
-                        batch_gen.add_fail(e.exit_code());
-                        continue;
+            'sector: for p in params.sectors {
+                let mut verified_infos = Vec::new();
+                let mut nonverified_deal_space: BigInt = BigInt::zero();
+                for deal_id in &p.deal_ids {
+                    // Check each deal is present only once, within and across sectors.
+                    if activated_deals.contains(deal_id) {
+                        log::warn!("failed to activate sector, duplicated deal {}", deal_id);
+                        batch_gen.add_fail(ExitCode::USR_ILLEGAL_ARGUMENT);
+                        continue 'sector;
                     }
-                };
 
-                let deal_spaces = match validate_and_return_deal_space(
-                    &proposals,
-                    &miner_addr,
-                    p.sector_expiry,
-                    curr_epoch,
-                    None,
-                ) {
-                    Ok(ds) => ds,
-                    Err(e) => {
-                        log::warn!("failed validate deals {:?}: {}", p.deal_ids, e);
-                        batch_gen.add_fail(e.exit_code());
-                        continue;
+                    let proposal = match preactivate_deal(
+                        rt,
+                        *deal_id,
+                        &proposals,
+                        &states,
+                        &pending_deals,
+                        &miner_addr,
+                        p.sector_expiry,
+                        curr_epoch,
+                        st.next_id,
+                    )? {
+                        Ok(v) => v,
+                        Err(e) => {
+                            log::warn!("failed to activate deal: {}", e);
+                            batch_gen.add_fail(e.exit_code());
+                            continue 'sector;
+                        }
+                    };
+
+                    // No continue below here, to ensure state changes are consistent.
+                    // Any error is an abort.
+                    // Extract and remove any verified allocation ID for the pending deal.
+                    let alloc_id = remove_pending_deal_allocation_id(
+                        // FIXME move outside inner loop
```

Comment:
I don't follow the fixme, this is a deal_id operation so doesn't it need to be inside this loop ?